feature/message_polling_backoff

BarAgam opened the pull request.

barmako committed:
Backoff mechanism implementations.
The integration isn't complete yet. as IntentService is incapable of rescheduling polling jobs in the future using a handler (since its thread is closing after its handling method is done).
A refactor needs to be made to change the message polling service into a ThreadHandler.

barmako committed:
Fully working exponential back-off message polling.

Message polling is now done in a separate  dedicated thread instead of using an intent service, hence the service java file has been deleted and its references across the application and the manifest file as-well.

Polling starts and ends just as before, with main activity active status (between onResume and onPause).

The repeatable action running that depends on some sort of backoff mechanism was written in its own class RepeatedBackoffTaskRunner.
It has its own implementation that instantiate it in a way that it uses a dedicated  thread (it is possible to make others TaskRunners that would use the same thread)
It uses an exponential backoff strategy and does the polling action via the IMessagePoller, made earlier (way-way earlier....)
The packages hierarchy changed a bit to fit the new state where no service is used for the message polling.

Auto-indentation across all "touched" files

review-ninja commented:
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/59" target="_blank"><img src="https://app.review.ninja/52010471/pull/59/badge" alt="ReviewNinja"/>

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/common/ExponentialBackoffStrategy.java:
```diff
@@ -0,0 +1,49 @@
+    private final int mMaxBackoffMillis;
+    private final int mBaseIntervalMillis;
+
+    private long mCurrentIntervalMillis;
```
why do you need both mCurrentIntervalMillis and mCurrentBackoffMillis? just multply each step with mMultiplier. you dont need to add mCurrentIntervalMillis.

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/common/ExponentialBackoffStrategy.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Override
+    public long getBackoffMillis() {
```
also the name Backoff is confusing. maybe getCurrentIntervalMillis(). or postponing ?

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/common/RepeatedBackoffTaskRunner.java:
```diff
@@ -0,0 +1,81 @@
+                    mBackoffStrategy.reset();
+                    onSuccessfulTask();
+                } catch (Exception ex) {
+                    mBackoffStrategy.increase();
```
increase wont always increase...

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/network/services/message_polling/RepeatedBackoffMessagePolling.java:
```diff
@@ -0,0 +1,121 @@
+    private static IMessagePoller createPoller(Context context) {
+        IMessageBroadcaster messageLocalBroadcaster = new MessageLocalBroadcaster(context);
+        IPolledMessagesProcessor processor = new PolledMessagesBroadcaster(messageLocalBroadcaster);
+        PreferenceUtil prefUtil = new PreferenceUtil(context.getResources(),
```
use ((GGApplication) context.getApplicationContext()).getPrefs()

BarAgam replied:
OK

BarAgam replied:
I also chnaged the signature of the creation method to accept only GGApplication to avoid bad-usage memory leaks

teamagam1 commented on app/src/test/java/com/teamagam/gimelgimel/app/common/ExponentialBackoffStrategyTest.java:
```diff
@@ -0,0 +1,79 @@
+
+        mExponentialBackoffStratgey.reset();
+
+        assertThat(mExponentialBackoffStratgey.getBackoffMillis(), equalTo(0L));
```
why getBackoffMillis can be 0L?

barmako committed:
Forced the creation of polling task via the application's context (to avoid memory leaks)
Used GGApplication's SecuredPreferenceUtil instead of creating my own PreferenceUtil

barmako committed:
Forced the creation of polling task via the application's context (to avoid memory leaks)
Used GGApplication's SecuredPreferenceUtil instead of creating my own PreferenceUtil.

renamed poller dir to polling.
